test/remote-github-patch-text.ts

```diff
@@ -0,0 +1,454 @@
+      .stub(logger, 'warn')
+      .resolves(listFilesOfPRResult as any);
+    sandbox
+      .stub(octokit.pulls, 'listFiles')
+      .resolves(listFilesOfPRResult as any);
+
+    // tests
+    const {filesMissingPatch} = await getCurrentPulLRequestPatches(
+      octokit,
+      upstream,
+      pullNumber,
+      pageSize
+    );
+    sandbox.assert.called(stub);
+    expect(filesMissingPatch.length).equals(1);
+    expect(filesMissingPatch[0]).equals(listFilesOfPRResult.data[0].filename);
+  });
+});
+
+describe('Extracting multiple GitHub files ranges from patch text', () => {
+  it('Returns an empty range file record when there is no patch text', () => {
+    const patchText = new Map<string, string>();
+    const ranges = patchTextToRanges(patchText);
+    expect(ranges.size).equals(0);
+  });
+  it('Throws an error when the patch text is an empty string', () => {
+    const patchText = new Map<string, string>();
+    patchText.set('invalid-patch.txt', '');
+    try {
+      patchTextToRanges(patchText);
+    } catch (err) {
+      expect(err.message).equals('Unexpected patch text format');
+    }
+  });
+  it('Throws an error when the patch text is a string that does not contain patch data', () => {
+    const patchText = new Map<string, string>();
+    patchText.set('invalid-patch.txt', '@@ this is some invalid patch data @@');
+    try {
+      patchTextToRanges(patchText);
+    } catch (err) {
+      expect(err.message).equals('Unexpected patch text format');
+    }
+  });
+  it('Calculates ranges with an inclusive lower bound and an exclusive upper bound', () => {
+    const multiline_patch =
+      '@@ -1,2 +1,5 @@\n Hello world\n-!\n+Goodbye World\n+gOodBYE world\n+\n+Goodbye World';
+    const multiline_file_name = 'multi-line-patch.txt';
+    const first_line_patch = "@@ -0,0 +1 @@\n+console.log('Hello World!');";
+    const milti_to_singleline_file_name = 'multi-to-single-line-patch.txt';
+    const patchText = new Map<string, string>();
+    patchText.set(multiline_file_name, multiline_patch);
+    patchText.set(milti_to_singleline_file_name, first_line_patch);
+    const ranges = patchTextToRanges(patchText);
+    expect(ranges.get(multiline_file_name)).not.equals(null);
+    expect(ranges.get(multiline_file_name)?.length).equals(1);
+    expect((ranges.get(multiline_file_name) as Range[])[0].start).equals(1);
+    expect((ranges.get(multiline_file_name) as Range[])[0].end).not.equals(5);
+    expect((ranges.get(multiline_file_name) as Range[])[0].end).equals(6);
+    expect(ranges.get(milti_to_singleline_file_name)).not.equals(null);
+    expect(ranges.get(milti_to_singleline_file_name)?.length).equals(1);
+    expect(
+      (ranges.get(milti_to_singleline_file_name) as Range[])[0].start
+    ).equals(1);
+    expect(
+      (ranges.get(milti_to_singleline_file_name) as Range[])[0].end
+    ).not.equals(1);
+    expect(
+      (ranges.get(milti_to_singleline_file_name) as Range[])[0].end
+    ).equals(2);
```

Comment:
These casts seem unnecessary if the typing is correct.

Reply:
The type is `undefined | Range[]` but it wouldn't let me do `ranges.get(milti_to_singleline_file_name)?[0].end)`